File: python/comm/data/DataCollection.py
from python.comm.data.CommunicationData import CommunicationData
from python.comm.data.utils import createCommunicationData
from python.comm.data.MessageType import MessageType
from typing import Dict


class DataCollection:
    def __init__(self):
        self.data = {}  # type: Dict[str, CommunicationData]

    def get(self, messageType: MessageType):
        communicationData = self.data.get(MessageType.messageTypeToString(messageType), None)
        if messageType != MessageType.NOTHING and communicationData is None:
            self.data[MessageType.messageTypeToString(messageType)] = createCommunicationData(messageType)
        return communicationData

File: csharp/Comm/data/DataCollection.cs
using System.Collections.Generic;

namespace Comm.data {
    public class DataCollection {

        public DataCollection() {
            this.data = new Dictionary<string, CommunicationData>();
        }

        ~DataCollection() {
            this.data.Clear();
        }

        public void set(MessageType messageType, CommunicationData commData) {
            string stringMessageType = MessageTypeConverter.messageTypeToString(messageType);
            this.data[stringMessageType] = commData;
        }

        public CommunicationData get(MessageType messageType) {
            CommunicationData commData;
            if (this.data.TryGetValue(MessageTypeConverter.messageTypeToString(messageType), out commData)) {
                return commData;
            }

            commData = DataUtils.createCommunicationData(messageType);
            this.data.Add(MessageTypeConverter.messageTypeToString(messageType), commData);
            return commData;
        }

        private Dictionary<string, CommunicationData> data;
    };
}
